File: src/python/examples/simple_inequality/driver.py
# In this example, we setup and minimize the problem
# min (x+1)^2+(y+1)^2 st x + 2y >= 1, 2x + y >=1

# Load in peopt
import peopt
from math import *

# Create a vector space 
class ListVS:
    def copy(self,x):
        return x
    def scal(self,alpha,x):
        return map(lambda y:alpha*y,x)
    def zero(self,x):
        return [0]*len(x)
    def axpy(self,alpha,x,y):
        return map(lambda z:alpha*z[0]+z[1],zip(x,y))
    def innr(self,x,y):
        return reduce(lambda x,y:x+y[0]*y[1],zip(x,y),0)
    def prod(self,x,y):
        return map(lambda z:z[0]*z[1],zip(x,y))
    def id(self,x):
        return [1]*len(x)
    def linv(self,x,y):
        return map(lambda z:z[1]/z[0],zip(x,y))
    def barr(self,x):
        return reduce(lambda x,y:x+log(y),x,0)
    def srch(self,x,y):
        alpha = float("inf")
        for i in xrange(0,len(x)):
            if x[i] < 0:
                alpha0 = -y[i]/x[i]
                if alpha0 < alpha:
                    alpha=alpha0
        return alpha
    def symm(self,x):
        return x 

# Define the objective 
# 
# f(x,y)=(x+1)^2+(y+1)^2
#
sq = lambda x:x*x
class QuadObj:
    # Evaluation of the Rosenbrock function
    def eval(self,x):
        return sq(x[0]+1)+sq(x[1]+1) 

    # Gradient
    def grad(self,x): 
        g=[0]*len(x)
        g[0]=2*(x[0]+1)
        g[1]=2*(x[1]+1)
        return g

    # Hessian-vector product
    def hessvec(self,x,dx):
        H_dx = [0]*len(x)
    	H_dx[0] = 2*dx[0] 
        H_dx[1] = 2*dx[1] 
        return H_dx

# Define the constraints 
#
# g(x,y)= [ x + 2y - 1] 
#         [ 2x + y - 1] 
#
class LinConst:
    # y=g(x) 
    def eval(self,x):
        y = [0]*2
        y[0]=x[0]+2*x[1]-1
        y[1]=2*x[0]+x[1]-1
        return y

    # y=g'(x)dx
    def p(self,x,dx):
        y = [0]*2
        y[0]= dx[0] + 2*dx[1]
        y[1]= 2*dx[0] + dx[1] 
        return y

    # z=g'(x)*dy
    def ps(self,x,dy):
        z = [0]*2
        z[0]= dy[0] + 2*dy[1] 
        z[1]= 2*dy[0] + dy[1]
        return z

    # z=(g''(x)dx)*dy
    def pps(self,x,dx,dy):
        z = [0]*2
        z[0] = 0.
        z[1] = 0. 
        return z

# Create a bundle of vector spaces 
class vs:
    X = ListVS()
    Z = ListVS()

# Create a bundle of functions
class fns:
    f = QuadObj()
    h = LinConst()

# Create a bundle of points
class pts:
    x = [3.,4.]
    dx = [5.,6.]
    dxx = [7.,8.]
    z = [1.,1.]
    dz = [9.,10.]

#peopt.diagnostics(vs(),fns(),pts())
sol = peopt.getMin(vs(),fns(),pts(),'simple_inequality.peopt')
print sol[0]

File: src/cpp/examples/simple_inequality/driver.cpp
// Optimize a simple optimization problem with an optimal solution
// of (1/3,1/3)

#include "optizelle/optizelle.h"
#include "optizelle/vspaces.h"
#include "optizelle/json.h"
#include <iostream>
#include <iomanip>
#include <cstdlib>

// Squares its input
template <typename Real>
Real sq(Real x){
    return x*x; 
}

// Define a simple objective where 
// 
// f(x,y)=(x+1)^2+(y+1)^2
//
struct MyObj
    : public Optizelle::ScalarValuedFunction <double,Optizelle::Rm>
{
    typedef Optizelle::Rm <double> X;

    // Evaluation 
    double operator () (const X::Vector& x) const {
        return sq(x[0]+1.)+sq(x[1]+1.);
    }

    // Gradient
    void grad(
        const X::Vector& x,
        X::Vector& grad
    ) const {
        grad[0]=2.*x[0]+2.;
        grad[1]=2.*x[1]+2.;
    }

    // Hessian-vector product
    void hessvec(
        const X::Vector& x,
        const X::Vector& dx,
        X::Vector& H_dx
    ) const {
        H_dx[0]=2.*dx[0]; 
        H_dx[1]=2.*dx[1]; 
    }
};

// Define simple inequalities 
//
// h(x,y)= [ x + 2y >= 1 ] 
//         [ 2x + y >= 1 ] 
//
struct MyIneq
    :public Optizelle::VectorValuedFunction<double,Optizelle::Rm,Optizelle::Rm>
{
    typedef Optizelle::Rm <double> X;
    typedef Optizelle::Rm <double> Y;

    // y=h(x) 
    void operator () (
        const X::Vector& x,
        Y::Vector& y
    ) const {
        y[0]=x[0]+2.*x[1]-1.;
        y[1]=2.*x[0]+x[1]-1.;
    }

    // y=h'(x)dx
    void p(
        const X::Vector& x,
        const X::Vector& dx,
        Y::Vector& y
    ) const {
        y[0]= dx[0]+2.*dx[1];
        y[1]= 2.*dx[0]+dx[1];
    }

    // z=h'(x)*dy
    void ps(
        const X::Vector& x,
        const Y::Vector& dy,
        X::Vector& z
    ) const {
        z[0]= dy[0]+2.*dy[1];
        z[1]= 2.*dy[0]+dy[1];
    }

    // z=(h''(x)dx)*dy
    void pps(
        const X::Vector& x,
        const X::Vector& dx,
        const Y::Vector& dy,
        X::Vector& z
    ) const {
        X::zero(z);
    }
};

int main(int argc,char* argv[]){
    // Read in the name for the input file
    if(argc!=2) {
        std::cerr << "simple_inequality <parameters>" << std::endl;
        exit(EXIT_FAILURE);
    }
    std::string fname(argv[1]);

    // Create a type shortcut
    using Optizelle::Rm;

    // Generate an initial guess
    std::vector <double> x(2);
    x[0]=2.1; x[1]=1.1;

    // Allocate memory for the inequality multipler 
    std::vector <double> z(2);

    // Create an optimization state
    Optizelle::InequalityConstrained <double,Rm,Rm>::State::t
        state(x,z);

    // Read the parameters from file
    Optizelle::json::InequalityConstrained <double,Optizelle::Rm,Optizelle::Rm>
        ::read(Optizelle::Messaging(),fname,state);
    
    // Create a bundle of functions
    Optizelle::InequalityConstrained <double,Rm,Rm>::Functions::t fns;
    fns.f.reset(new MyObj);
    fns.h.reset(new MyIneq);

    // Solve the optimization problem
    Optizelle::InequalityConstrained <double,Rm,Rm>::Algorithms
        ::getMin(Optizelle::Messaging(),fns,state);

    // Print out the reason for convergence
    std::cout << "The algorithm converged due to: " <<
        Optizelle::StoppingCondition::to_string(state.opt_stop) <<
        std::endl;

    // Print out the final answer
    const std::vector <double>& opt_x=*(state.x.begin());
    std::cout << std::scientific << std::setprecision(16)
        << "The optimal point is: (" << opt_x[0] << ','
	<< opt_x[1] << ')' << std::endl;

    // Write out the final answer to file
    Optizelle::json::InequalityConstrained<double,Rm,Rm>
        ::write_restart(Optizelle::Messaging(),"solution.json",state);

    // Return that the program exited properly
    return EXIT_SUCCESS;
}
